### ISSUE TITLE
mvcc.store.restore taking too long triggers snapshot cycle

### ISSUE THREAD

**Issue by gyuho** [coreos/etcd#5317]
#### Problem

functional-tester etcd-tester's `setHealthKey` gets `context timeout deadline exceeded`, while etcd servers are getting `etcdserver: publish error: etcdserver: request timed out`.

<br>

#### How to reproduce
1. Set up 3 node cluster
2. Write 8M keys
3. Kill one of the follower
4. In parallel, write more keys to other two members _(10K~15K QPS)_
5. Wait until +10K more keys are written (just enough to trigger snapshot)
6. Recover the follower
7. Then the follower reads the `db` file from the disk, and rebuilds index of 8M keys, by calling kvstore.Restore
8. Follower is now available again (able to talk to leader)
9. Follower fell behind as much the entries as were written in the leader while the follower was down
10. Leader sends snapshot to this follower
11. Follower receives this snapshot, which requires to rebuild index of 8M + 10K keys
12. While follower restores from this snapshot, it returns 'etcdserver: publish error: etcdserver:'
13. Stresser has been writing to leader in parallel, and it triggers another snapshot
14. Repeats from step 10.

It will be either compaction timeout or recovery timeout in functional-tests.

For example, use this benchmark command to stress etcd cluster:

```
benchmark --endpoints=${HOST_1}:2379,${HOST_2}:2379,${HOST_3}:2379 \
    --conns=100 --clients=1000 \
    put --key-size=10 --val-size=1 --total=3000000
```

And when the snapshot is triggered, one can see these in leader's logs:

```
2016-06-07 17:29:32.776501 I | etcdserver: wrote database snapshot out [total bytes: 235487232]
2016-06-07 17:29:33.649890 I | rafthttp: database snapshot [index: 3552177, to: 5fbb5cbbbef242b7] sent out successfully
```

And follower logs will be something like:

```
2016-06-07 17:29:33.649748 I | snap: saved database snapshot to disk [total bytes: 235487232]
2016-06-07 17:29:33.649800 I | rafthttp: received and saved database snapshot [index: 3552177, from: b88fca7fa2ea1289] successfully
2016-06-07 17:29:33.650026 I | raft: 5fbb5cbbbef242b7 [commit: 3000778, lastindex: 3000778, lastterm: 5] starts to restore snapshot [index: 3552177, term: 5]
2016-06-07 17:29:33.650089 I | raft: log [committed=3000778, applied=3000778, unstable.offset=3000779, len(unstable.Entries)=0] starts to restore snapshot [index: 3552177, term: 5]
2016-06-07 17:29:33.650129 I | raft: 5fbb5cbbbef242b7 restored progress of 5fbb5cbbbef242b7 [next = 3552178, match = 3552177, state = ProgressStateProbe, waiting = false, pendingSnapshot = 0]
2016-06-07 17:29:33.650148 I | raft: 5fbb5cbbbef242b7 restored progress of b88fca7fa2ea1289 [next = 3552178, match = 0, state = ProgressStateProbe, waiting = false, pendingSnapshot = 0]
2016-06-07 17:29:33.650163 I | raft: 5fbb5cbbbef242b7 restored progress of c4268f53965ad024 [next = 3552178, match = 0, state = ProgressStateProbe, waiting = false, pendingSnapshot = 0]
2016-06-07 17:29:33.650179 I | raft: 5fbb5cbbbef242b7 [commit: 3552177] restored snapshot [index: 3552177, term: 5]
2016-06-07 17:29:33.650323 I | etcdserver: applying snapshot at index 3000641...
2016-06-07 17:29:33.653349 I | etcdserver: raft applied incoming snapshot at index 3552177
2016-06-07 17:29:33.664600 I | etcdserver: restoring mvcc store...
2016-06-07 17:29:57.641605 I | etcdserver: finished restoring mvcc store
...
```

<br>

#### Root Cause

The root cause was the writes being too intensive, with the follower keep falling behind. When an etcd node recovers, it needs to rebuild the key-index from disk. Reading the backend file itself doesn't take long. Iterating 10M keys in boltdb takes only about 3 seconds in SSD.

Rebuilding the key-index from scratch takes long time.

Restoring 8M keys from scratch can take more than 30 seconds. And if there were concurrent writes _(10K~15K QPS)_ in other nodes, the recovering node will need to receive snapshot right after key-index rebuild, and the snapshot will trigger another index rebuilding. And this cycle repeats, as long as there are ongoing concurrent writes, with restoring keys taking longer and longer.

<br>

#### Conclusion

This only happens with extremely intense writes (large number of keys to rebuild btree). And we believe that this will rarely happen in production. Even with this high workload, etcd leader kept operating. If anyone is hitting this issue, please check your workload, try slowing down the workload, or consider compacting your database.

**There was nothing wrong in etcd correctness.**

We mark this as unplanned.

In the future, we will make the `restore` operation faster.

<br>

#### More details

Did more timing:

```
var kv mvccpb.KeyValue
st1 := time.Now()
if err := kv.Unmarshal(vals[i]); err != nil {
    plog.Fatalf("cannot unmarshal event: %v", err)
}
dTime += time.Since(st1)

st2 := time.Now()
s.kvindex.Restore(kv.Key, revision{kv.CreateRevision, 0}, rev, kv.Version)
rTime += time.Since(st2)

mvcc: [DEBUG]
store.restore took 36.38789088s
decode took 2.106427628s
kvindex.restore took 31.230108334s
```

So it's not the decode that take much time, it's `kvindex.Restore` that takes most of the time.

<br>
/cc @heyitsanthony @xiang90

**Comment by gyuho** [comment]
Closing. Found out it was actually timing out from compact request, which I am still investigating.

**Comment by gyuho** [comment]
```
2016-06-03 06:50:28.610202 I | etcd-tester: [round#1604 case#2] succeed!
2016-06-03 06:50:28.652412 I | etcd-tester: [round#1604 case#3] injecting failure "kill leader member"
2016-06-03 06:50:28.737285 I | etcd-tester: [round#1604 case#3] injected failure
2016-06-03 06:50:28.737333 I | etcd-tester: [round#1604 case#3] recovering failure "kill leader member"
2016-06-03 06:50:29.738835 W | etcd-tester: #0 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.2:2379))
2016-06-03 06:50:31.752794 W | etcd-tester: #1 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:50:33.761561 W | etcd-tester: #2 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:50:35.773781 W | etcd-tester: #3 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:50:37.790226 W | etcd-tester: #4 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:50:39.804781 W | etcd-tester: #5 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:50:41.818040 W | etcd-tester: #6 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:50:43.830512 W | etcd-tester: #7 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:50:45.845397 W | etcd-tester: #8 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:50:47.863444 W | etcd-tester: #9 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:50:49.876040 W | etcd-tester: #10 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:50:51.890448 W | etcd-tester: #11 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:50:53.913856 W | etcd-tester: #12 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
...
2016-06-03 06:52:06.803569 W | etcd-tester: #48 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:52:08.823311 W | etcd-tester: #49 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:52:10.930699 W | etcd-tester: #50 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:52:12.945202 W | etcd-tester: #51 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:52:13.945370 I | etcd-tester: [DEBUG] setHealthKey taking too long...
2016-06-03 06:52:14.960676 W | etcd-tester: #52 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:52:15.960864 I | etcd-tester: [DEBUG] setHealthKey taking too long...
...
2016-06-03 06:53:56.948635 I | etcd-tester: [DEBUG] setHealthKey taking too long...
2016-06-03 06:53:57.970383 W | etcd-tester: #103 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:53:58.970540 I | etcd-tester: [DEBUG] setHealthKey taking too long...
2016-06-03 06:53:59.984046 W | etcd-tester: #104 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.3:2379))
2016-06-03 06:54:00.984188 I | etcd-tester: [DEBUG] setHealthKey taking too long...
...
2016-06-03 06:57:53.509225 I | etcd-tester: [DEBUG] setHealthKey taking too long...
2016-06-03 06:57:53.513402 W | etcd-tester: #299 setHealthKey error (rpc error: code = 8 desc = etcdserver: mvcc: database space exceeded (10.240.0.2:2379))
2016-06-03 06:57:54.513683 I | etcd-tester: [DEBUG] setHealthKey taking too long...
2016-06-03 06:57:54.513770 I | etcd-tester: [round#1604 case#3] recovery error: rpc error: code = 8 desc = etcdserver: mvcc: database space exceeded (10.240.0.2:2379)
2016-06-03 06:57:54.513794 I | etcd-tester: [round#1604 case#3] cleaning up...
2016-06-03 06:57:57.399367 W | etcd-tester: #0 setHealthKey error (rpc error: code = 4 desc = context deadline exceeded (10.240.0.2:2379))
2016-06-03 06:57:58.416503 I | etcd-tester: [round#1604 case#0] injecting failure "kill all members"
2016-06-03 06:57:58.436832 I | etcd-tester: [round#1604 case#0] injected failure
```

So agent2(`10.240.0.3:2379`) is taking long...

```
# agent2

2016-06-03 06:50:28.763581 I | etcdmain: etcd Version: 3.0.0-beta.0+git
2016-06-03 06:50:28.763633 I | etcdmain: Git SHA: 17f43ec
2016-06-03 06:50:28.763640 I | etcdmain: Go Version: devel +e29e0ba Wed Jun 1 13:33:43 2016 +0000
2016-06-03 06:50:28.763645 I | etcdmain: Go OS/Arch: linux/amd64
2016-06-03 06:50:28.763651 I | etcdmain: setting maximum number of CPUs to 4, total number of available CPUs is 4
2016-06-03 06:50:28.763695 N | etcdmain: the server is already initialized as member before, starting as etcd member...
2016-06-03 06:50:28.763743 I | etcdmain: listening for peers on http://10.240.0.3:2380
2016-06-03 06:50:28.763778 I | etcdmain: listening for client requests on 10.240.0.3:2379
2016-06-03 06:50:28.818899 W | snap: skipped unexpected non snapshot file 000000002db6971b.snap.db
2016-06-03 06:50:28.818943 W | snap: skipped unexpected non snapshot file 00000000436b75b1.snap.db
2016-06-03 06:50:28.818950 W | snap: skipped unexpected non snapshot file 0000000021155849.snap.db
2016-06-03 06:50:28.818955 W | snap: skipped unexpected non snapshot file 0000000017cd62d2.snap.db
2016-06-03 06:50:28.818972 W | snap: skipped unexpected non snapshot file 00000000436b961a.snap.db
2016-06-03 06:50:28.831224 I | etcdserver: recovered store from snapshot at index 1374613998
2016-06-03 06:50:28.831241 I | etcdserver: name = etcd-1
2016-06-03 06:50:28.831247 I | etcdserver: data dir = agent.etcd
2016-06-03 06:50:28.831253 I | etcdserver: member dir = agent.etcd/member
2016-06-03 06:50:28.831258 I | etcdserver: heartbeat = 100ms
2016-06-03 06:50:28.831265 I | etcdserver: election = 1000ms
2016-06-03 06:50:28.831271 I | etcdserver: snapshot count = 10000
2016-06-03 06:50:28.831282 I | etcdserver: advertise client URLs = http://10.240.0.3:2379
2016-06-03 06:50:28.910081 I | etcdserver: restarting member ab5aadb7a621058 in cluster 4bb6592ff0063e99 at commit index 1374614997
2016-06-03 06:50:28.910267 I | raft: ab5aadb7a621058 became follower at term 9275
2016-06-03 06:50:28.910288 I | raft: newRaft ab5aadb7a621058 [peers: [ab5aadb7a621058,6bdc140107d9d490,f9aff54dd2d6503e], term: 9275, commit: 1374614997, applied: 1374613998, lastindex: 1374615119, lastterm: 9275]
2016-06-03 06:50:28.910987 I | membership: added member ab5aadb7a621058 [http://10.240.0.3:2380] to cluster 4bb6592ff0063e99 from store
2016-06-03 06:50:28.911000 I | membership: added member f9aff54dd2d6503e [http://10.240.0.2:2380] to cluster 4bb6592ff0063e99 from store
2016-06-03 06:50:28.911007 I | membership: added member 6bdc140107d9d490 [http://10.240.0.4:2380] to cluster 4bb6592ff0063e99 from store
2016-06-03 06:50:28.911014 I | membership: set the cluster version to 3.0 from store
2016-06-03 06:50:28.914186 I | mvcc: restore compact to 1367234531
2016-06-03 06:50:52.262878 I | rafthttp: starting peer 6bdc140107d9d490...
2016-06-03 06:50:52.262924 I | rafthttp: started HTTP pipelining with peer 6bdc140107d9d490
2016-06-03 06:50:52.263279 I | rafthttp: started streaming with peer 6bdc140107d9d490 (writer)
2016-06-03 06:50:52.263346 I | rafthttp: started streaming with peer 6bdc140107d9d490 (writer)
```

**Comment by gyuho** [comment]
Problem
- Follower was taking too long to recover(restart) (+30 seconds)
- In the meantime, leader uses all 30-second buffer to send another snapshot
- Follower keeps falling behind

I will try to reproduce locally by forcing the cluster leader to send snapshot with slow follower to get the exact numbers that can cause this problem. And will investigate more how the follower takes +30 seconds to recover (boltdb, number of keys, size of data).

**Comment by xiang90** [comment]
Ideally we want etcd to recover from 40M small keys within 30 seconds.

40M is around 1 hour 10k keys/second put workload without any compaction. This is probably the max throughput we want to support for etcd.

/cc @nekto0n @sinsharat @vimalk78 It would be great if any of you can work on it.

**Comment by nekto0n** [comment]
According to @gyuho most of the time is taken by `kvindex.Restore`. There is a `BenchmarkIndexRestore` which on my laptop gives me these numbers:
```
1000000	      1411 ns/op	     178 B/op	       3 allocs/op
```
Using only 100k keys (simulating multiple revisions) I get:
```
2000000	       687 ns/op	     119 B/op	       1 allocs/op
```
So it takes a bit longer than a second to perform 1M keys recovery. Am I right that we need to optimize this a bit more or should we concoct a more involving test which should look like real code [here](https://github.com/coreos/etcd/blob/master/mvcc/kvstore.go#L390)?

**Comment by xiang90** [comment]
> So it takes a bit longer than a second to perform 1M keys recovery and a bit less for 100k keys with total revisions number of 2M. 

Is it growing linearly? 100K -> 10 rev pers key = 1s (1M keys in total); 100k -> 10 revs = 2s (1M keys in total)?

I remembered I did some experiments a while ago and found some low hanging fruits.

**Comment by nekto0n** [comment]
I did [a bit of experimenting with benchmark](https://github.com/nekto0n/etcd/blob/d73eec82f03929508333e591b00501a22e1989d6/mvcc/index_bench_test.go) and got these numbers:
`go test -v github.com/coreos/etcd/mvcc -bench BenchmarkIndexRestore -run ^$ -benchtime 10s`

```
BenchmarkIndexRestore1-4     	10000000	      1314 ns/op	     178 B/op	       3 allocs/op
BenchmarkIndexRestore10-4    	20000000	       737 ns/op	     123 B/op	       1 allocs/op
BenchmarkIndexRestore100-4   	30000000	       624 ns/op	     105 B/op	       1 allocs/op
```

>I remembered I did some experiments a while ago and found some low hanging fruits.

Can you share some directions where they may be hanging from? :)

**Comment by xiang90** [comment]
> Can you share some directions where they may be hanging from? :)

Sure. During the restore, we keep on touching the treeIndex for single key get and single key put. It does one get, put for every key and every key modification. These operations for tree are O(NlogN), which is not ideal.

To improve this, we can have a unordered map to store the temp data. We only touch the unordered map while restoring the index. After we iterate the all the revision, we range over the unordered map and convert it to the tree index. This can reduce the cost significantly if we have N versions per key.

Some simple code:

```go
        // use an unordered map to hold the temp index data to speed up
	// the initial key index recovery.
	// we will convert this unordered map into the tree index later.
	unordered := make(map[string]*keyIndex, 100000)

	// TODO: limit N to reduce max memory usage
	keys, vals := tx.UnsafeRange(keyBucketName, min, max, 0)
	for i, key := range keys {
		var kv mvccpb.KeyValue
		if err := kv.Unmarshal(vals[i]); err != nil {
			plog.Fatalf("cannot unmarshal event: %v", err)
		}

		rev := bytesToRev(key[:revBytesLen])

		is := time.Now()
		// restore index
		switch {
		case isTombstone(key):
			if ki, ok := unordered[string(kv.Key)]; ok {
				ki.tombstone(rev.main, rev.sub)
			}
			delete(keyToLease, string(kv.Key))

		default:
			ki, ok := unordered[string(kv.Key)]
			if ok {
				ki.put(rev.main, rev.sub)
			} else {
				ki = &keyIndex{key: key}
				ki.restore(revision{kv.CreateRevision, 0}, rev, kv.Version)
				unordered[string(kv.Key)] = ki
			}

			if lid := lease.LeaseID(kv.Lease); lid != lease.NoLease {
				keyToLease[string(kv.Key)] = lid
			} else {
				delete(keyToLease, string(kv.Key))
			}
		}

		// update revision
		s.currentRev = rev
	}

	// restore the tree index from the unordered index.
	for _, v := range unordered {
		s.kvindex.Insert(v)
	}
```

```go
func (ti *treeIndex) Insert(ki *keyIndex) {
	ti.Lock()
	defer ti.Unlock()
	ti.tree.ReplaceOrInsert(ki)
}
```

On my laptop, it reduces the 10M revisions (1M keys and 10 versions pre key) recovery time from 40s to 10s.

For the unordered map recovery, we can make it even fancier by translating it into map-reduce style and increase concurrency. But that is a future step and we might not even need it.

**Comment by nekto0n** [comment]
So we change `O(nlogn)` (for every key version we use 2 get/put to the tree) with `O(n) + O(logn)` (every key is put into hashtable and then about 1/10 of them are put into the tree).
All is needed is to copy-paste the code from above into a PR, sounds doable :)

**Comment by xiang90** [comment]
@nekto0n Some tests need to be changed. Benchmark needs to be updated or added.

**Comment by sinsharat** [comment]
@nekto0n are you going to add the code as well as benchmark or can i take any part of it?

**Comment by nekto0n** [comment]
@sinsharat actually I'm pretty busy this week, so feel free to take all of it.

**Comment by sinsharat** [comment]
@nekto0n Sure will start looking into it then.
Thanks!

**Comment by sinsharat** [comment]
@xiang90 sorry for the delay on this. Was stuck with some business commitments, which i was asked to work on this entire week. Will try to finish this and raise a PR for this by tomorrow.
Thank you!

**Comment by xiang90** [comment]
Fixed by #6846. There are a few other places that can be improved. But we will optimize it until it becomes a problem in real world.
